baseplate/frameworks/queue_producer/avro_kafka.py

```diff
@@ -0,0 +1,97 @@
+import logging
+
+from typing import Any
+from typing import Dict
+from typing import Optional
+
+from confluent_kafka.avro import AvroProducer
+
+from baseplate import _ExcInfo
+from baseplate import Span
+from baseplate import SpanObserver
+from baseplate.clients import ContextFactory
+from baseplate.lib import config
+
+logger = logging.getLogger(__name__)
+
+
+class AvroProducerContextFactory(ContextFactory):
+    def __init__(self, bootstrap_servers: str, schema_registry: str, acks: int):
+        self.bootstrap_servers = bootstrap_servers
+        self.schema_registry = schema_registry
+        self.acks = acks
+
+    def make_object_for_context(self, name: str, span: Span) -> AvroProducer:
+        avro_producer = KafkaAvroProducer(
+            name,
+            span,
+            {
+                "bootstrap.servers": self.bootstrap_servers,
+                "schema.registry.url": self.schema_registry,
+                "acks": self.acks,
+            },
+        )
+        span.register(AvroProducerSpanObserver(avro_producer))
+        return avro_producer
+
+
+class KafkaAvroProducer:
+    def __init__(self, name: str, span: Span, app_config: Dict[str, Any]):
+        self.name = name
+        self.span = span
+        self.avro_producer = AvroProducer(app_config)
+
+    def produce(self, topic: str, schema_id: int, value: Dict[str, Any]) -> None:
+        """Encode `value` using the `schema_id` and produce the message to Kafka.
+
+        :param topic: topic name
+        :param schema_id: schema id associated with the schema that `value` conforms to. It will
+                          be used to look up the schema from Schema Registry. This prevents callers
+                          from using unregistered schemas
+       :param value: object to serialize
+
+        """
+        serializer = self.avro_producer._serializer
+
+        encode_trace_name = "{}.{}".format(self.name, "serializer.encode_record_with_schema_id")
+        encode_span = self.span.make_child(encode_trace_name)
+
+        with encode_span:
+            buffer = serializer.encode_record_with_schema_id(schema_id, value)
+
+        producer_trace_name = "{}.{}".format(self.name, "avro_producer.produce")
+        producer_span = self.span.make_child(producer_trace_name)
+
+        producer = super(AvroProducer, self.avro_producer)  # pylint: disable=bad-super-call;
+        with producer_span:
+            producer.produce(topic=topic, value=buffer)  # type: ignore
+
+
+class AvroProducerSpanObserver(SpanObserver):
+    """Automatically flush to kafka at the end of each request."""
+
+    def __init__(self, kafka_avro_producer: KafkaAvroProducer):
+        self.avro_producer = kafka_avro_producer.avro_producer
+
+    def on_finish(self, exc_info: Optional[_ExcInfo]) -> None:
+        self.avro_producer.flush()
```

Comment:
Why are we flushing the producer at the end of the request rather than at the call time?  That seems like something we wouldn't want to do since you wouldn't be able to rely on on knowing that your message was actually published or not when making the function call.

Reply:
I think the thinking is that we can let the background publishing do its thing and not block, but we only want to be sure the message was produced by the time we are exiting

Reply:
Right, but why do that?  We don't do that for any other IO clients like this and it means that the caller now no longer knows for sure that its message was actually published to Kafka.  If the app crashes or there's a network failure before this happens, we might have told the caller that their request succeeded but their message did not actually make it from Kafka, so they don't even get the chance to retry and that message has been lost.

Reply:
So do you think we should depend on the client calling `flush()`? or build that into our wrapper around the real avro client?

Reply:
Even if we go with this offline publishing approach, I don't think we can use `flush` since it'll block gevent:

https://github.com/confluentinc/confluent-kafka-python/issues/612

Reply:
I think we should build calling `poll(timeout=0)` into the client.  You can pass in a callback to the produce function that'll get called after the message is delivered, one potential option would be something like this:

```
def produce(message):
    q = queue.Queue(maxsize=1)
    func on_delivery(message, err):
        q.add((message, err))
    kafka.publish(message, on_delivery=on_delivery)
    message, err = flush(q)
    return message, err

def _flush(q):
    kafka.poll(timeout=0)
    while q.empty():
        kafka.poll(timeout=0)
    return q.get(block=False)
```

Reply:
why use a queue instead of calling `poll` or `flush` in a loop with 0 timeout and small sleep and then waiting until the number of pending messages is 0 https://docs.confluent.io/current/clients/confluent-kafka-python/index.html#confluent_kafka.Producer.flush?


something like:

```
def produce(...):
    kafka.produce(topic, value)

def on_finish(...):
    pending_messages = self.avro_producer.flush(timeout=0)
    while pending_messages:
            # we can't use timeouts on the kafka consumer methods
            # because they call out to C code that ends up blocking
            # the event loop. gevent will successfully yield control
            # on this sleep call though.
            # see https://tech.wayfair.com/2018/07/blocking-io-in-gunicorn-gevent-workers/
            logger.debug("waiting 0.001s for pending messages to flush")
            time.sleep(0.001)
            pending_messages = self.avro_producer.flush(timeout=0)
```

Reply:
I think there's 2 things, one is that we probably don't want the flush logic to live in `on_finish`, I still think we want to flush within the `produce` call so the client using the `produce` method can know that their message has been delivered.

Now, about how we accomplish that, the queue _might_ be overkill but it does let us know that our message was delivered, regardless if any other pending messages were delivered.  If we want to just spin until `pending` is 0, we could be waiting an indeterminate amount of time for other messages to publish.

Reply:
Actually, I would think we would still want the queue (or something like it), since that seems to be the only way to actually get the message from the `on_delivery` callback into the main call stack to return up to the caller of `publish` (or raise `err` possibly?).

Reply:
Ok how about having the flush in `on_finish` but also providing a "safer" producer message that ensures the message is actually sent?

Reply:
OK how about something like https://gist.github.snooguts.net/bsimpson/7a63c0da25075781d873a6843460a853 using thread locals to get the result back to the caller?

Reply:
why is it important to be in the `on_finish` call instead of at the time `produce` is called? if you keep the `flush` in `on_finish` aren't you enabling developers to write unexpectedly slow/blocking producers?

Reply:
Thread local approach looks good!

Reply:
@manishapme yeah I think for now we can drop the `on_finish` and wait for the response in our `produce()` method